packages/cli/src/index.js

```diff
@@ -31,69 +32,117 @@ import * as typeCheckCommand from './commands/type-check'
-      delete argv.cwd
-    } else if (process.env.RWJS_CWD) {
-      cwd = process.env.RWJS_CWD
-    } else {
-      cwd = path.dirname(getConfigPath())
+// # Setting the CWD
+//
+// The current working directory can be set via:
+//
+// 1. The `--cwd` option
+// 2. The `RWJS_CWD` env-var
+// 3. By traversing directories upwards for the first `redwood.toml`
+//
+// ## Examples
+//
+// ```
+// yarn rw info --cwd /path/to/project
+// RWJS_CWD=/path/to/project yarn rw info
+//
+// # In this case, `--cwd` wins out over `RWJS_CWD`
+// RWJS_CWD=/path/to/project yarn rw info --cwd /path/to/other/project
+//
+// # Here we traverses upwards for a redwood.toml.
+// cd api
+// yarn rw info
+// ```
+
+let { cwd, testEnv } = Parser(hideBin(process.argv))
+cwd ??= process.env.RWJS_CWD
+
+try {
+  if (cwd) {
+    // `cwd` was set by the `--cwd` option or the `RWJS_CWD` env var. In this case,
+    // we don't want to find up for a `redwood.toml` file. The `redwood.toml` should just be in that directory.
+
+    const redwoodTOMLPath = path.join(cwd, 'redwood.toml')
+
+    if (!fs.existsSync(redwoodTOMLPath)) {
+      throw new Error(`Couldn't find a "redwood.toml" file in ${cwd}.`)
     }
+  } else {
+    // `cwd` wasn't set. Odds are they're in a Redwood project,
+    // but they could be in ./api or ./web, so we have to find up to be sure.
+
+    const redwoodTomlPath = findup('redwood.toml', { cwd: process.cwd() })
 
-    configPath = path.resolve(process.cwd(), cwd, 'redwood.toml')
-    if (!fs.existsSync(configPath)) {
-      throw new Error('Could not find `redwood.toml` config file.')
+    if (!redwoodTomlPath) {
+      throw new Error(
+        `Couldn't find up a "redwood.toml" file from ${process.cwd()}.`
+      )
     }
 
-    process.env.RWJS_CWD = cwd
-  } catch (e) {
-    console.error()
-    console.error('Error: Redwood CLI could not find your config file.')
-    console.error(`Expected '${configPath}'`)
-    console.error()
-    console.error(`Did you run Redwood CLI in a RedwoodJS project?`)
-    console.error(`Or specify an incorrect '--cwd' option?`)
-    console.error()
-    process.exit(1)
+    cwd = path.dirname(redwoodTomlPath)
   }
+} catch (error) {
+  console.error(
+    [error.message, "Are you sure you're in a Redwood project?"].join('\n')
+  )
+  process.exit(1)
 }
 
-const loadDotEnvDefaultsMiddleware = () => {
-  config({
-    path: path.join(getPaths().base, '.env'),
-    defaults: path.join(getPaths().base, '.env.defaults'),
-    multiline: true,
-  })
+process.env.RWJS_CWD = cwd
+
+// # Load .env, .env.defaults
+//
+// This should be done as early as possible, and the earliest we can do it is after setting `cwd`.
+
+if (testEnv) {
+  console.log(
+    JSON.stringify({
+      name: 'process.env before dotenv-defaults',
+      value: process.env,
+    })
+  )
+}
+
+config({
+  path: path.join(getPaths().base, '.env'),
+  defaults: path.join(getPaths().base, '.env.defaults'),
+  multiline: true,
+})
+
+if (testEnv) {
```

Comment:
Instead of having a test-specific hidden cli option, plus test specific code here in the cli index file, could you move this code into a rw exec script and run that?

It wouldn't be able to test the "before" values, but it should be able to test the actual values that will be available to a Redwood app. And in a more app-like environment too.

Reply:
Yeah I don't love having test code here, and the exec script is a good idea but I'm not sure it would test the case I want. (Honestly this code here isn't testing the case I want either, so let me get that part straight.)

What I really want to test is that redwood calls dotenv (technically dotenv-defaults) first. The only other thing I know that was calling it before redwood was prisma.

I feel like I should be able to use jest to mock or "spy on" dotenv and check what calls it in what order, but I couldn't get the mock working.

Reply:
> What I really want to test is that redwood calls dotenv (technically dotenv-defaults) first. 

First, as in before Prisma? That's a pretty specific test case. Can you make the problem statement more generic? Maybe that'll make it easier to figure out how/what to test

Reply:
I prefer Dom's method here, because the more layers you have inbetween (exec is fairly complicated, with babel hooks and all) - the less likely you are to find the real problem.

One way to get around the test specific code, is to have a flag `--print-env-vars` - which may be useful to users too.

Reply:
I've removed the --test-env flag for now. I don't have a great way of testing it yet